Feature: Add first version of openAPI3 definition

antonykamp opened the pull request:
Fixes #52

This PR implements the first version of the openAPI3 definition.
It occurs that this contains more changes than I thought 🎃 
My suggestion is the following: Everybody reviews their component. For example, @CR1337 reviews everything connected to the `/component/spawner` endpoint. (Sorry, @instant-sky 😇). And somebody reviews Lucas' part and the index file, please 😄 
OpenAPI3 is very readable, IMO, and it's just a configuration with no functionality. You will rock this 🚀 

You can find some information about OpenAPI 3.0 and the file structure in the wiki: https://github.com/BP2022-AP1/bp2022-ap1/wiki/REST-API#api-definition-openapi-30

## PR checklist

- [x] Acceptance criteria fulfilled
- [x] Additional features are tested
- [x] Docs (code, wiki & diagrams) updated
- [x] Breaking changes anounced
- [ ] Dev-branch has been merged into local branch to resolve conflicts
- [ ] Tests and linter have passed AFTER local merge
- [ ] Another dev reviewed and approved

antonykamp committed:
feat: add first version of openAPI3 definition

antonykamp committed:
Merge branch 'dev' into feature/52-implement-first-version-of-rest-api-definition

SaturnHafen commented on postman/schemas/schemas/DataScience.yaml:
```diff
@@ -0,0 +1,8 @@
+components:
+  schemas:
+    DataScienceConfiguration:
+      type: object
+      properties: {}
+    DataScienceResult:
+      type: object
+      properties: {}
```
It is intended that this type is not specified any further, right?

antonykamp replied:
Yes 👍 I don't want to add content to the schemas at the moment, because this will happen in other issues 😇

SaturnHafen commented on postman/schemas/paths/token.yaml:
```diff
@@ -0,0 +1,41 @@
+# /token
```
**One does not store credentials in cleartext** - so this should return either friendly-names or only the first/last ~4 characters of the token. The database should only contain hashes of the tokens (and a mapping between the friendly name/partial token and the hashes).
Verification should be performed by hashing the token and comparing it to the database

antonykamp replied:
You're absolutely right and thanks for the reminder 👍 Do you think we need this route? Maybe we should only provide a route for generating a new token first, shouldn't we? 🤔

SaturnHafen commented on postman/schemas/paths/token.yaml:
```diff
@@ -0,0 +1,41 @@
+            description: API token missing
+        "403":
+            description: No rights to see the requested content
+put:
+    summary: Create new token with specified permission specified in body if current user is admin
+    operationId: createToken
+    requestBody:
+        required: true
+        content:
+            application/json:
+                schema:
+                    type: object
+                    properties:
+                        permission:
+                            type: string
+    responses:
+        "201":
+            description: Successfully created new token
+            content:
+                application/json:
+                    schema:
+                        $ref: "../schemas/Token.yaml#/components/schema/Token"
```
The token itself ~~must~~ should be generated on the server to make sure that the token has a high-enough entropy to be secure

antonykamp replied:
Thanks for the reminder 👍 I'll keep that in mind and I hope I can request a review from you when I implement this section 😇

SaturnHafen commented on postman/schemas/paths/token-by-id.yaml:
```diff
@@ -0,0 +1,45 @@
+        "404":
+            description: Id not found
+delete:
+    summary: Remove token if current user is admin.
```
you should be able to remove the token that you are currently using, shouldn't you? :thinking:

antonykamp replied:
I'm currently thinking about removing the possibility to remove a token at first.

SaturnHafen commented on postman/schemas/paths/component/data-science.yaml:
```diff
@@ -0,0 +1,50 @@
+                                type: string
+                                format: uuid
+        "400":
+            description: An data-science configuration exists for this simulation already
```
Is the data-science component "single-configuration"? i.e. the component can only have one configuration? I guess that makes sense...

antonykamp replied:
I'll keep it "single-configurated" as long as Lucas is on vacation 👍

SaturnHafen commented on postman/schemas/paths/component/data-science.yaml:
```diff
@@ -0,0 +1,50 @@
+          schema:
+              type: string
+              format: uuid
+          description: Specify id of simulation if you only want to get the data-science configuration of a single simulation
```
I don't know if you can link to an other route but you should at least reference the GET to `/component/data-science/{id}`

antonykamp replied:
Great idea 👍 We could use [Links](https://swagger.io/docs/specification/links/) but they are used in responses to link to follow-up requests. I'll create an issue to improve the API documentation 👍 see #125

SaturnHafen commented on postman/schemas/paths/component/data-science-by-id.yaml:
```diff
@@ -0,0 +1,64 @@
+# /component/data-science/{id}
```
Why are we sometimes using the query parameter and sometimes the path?

antonykamp replied:
I remember, that it's quite common to use path parameters to reference an object of the class (in this case data-science) and query parameters to specify the environment around. 😇 
See https://stackoverflow.com/questions/30967822/when-do-i-use-path-params-vs-query-params-in-a-restful-api for example ;)

SaturnHafen commented on postman/schemas/schemas/Run.yaml:
```diff
@@ -0,0 +1,8 @@
+    RunConfiguraion:
+      type: object
+      properties: {}
+    RunStatus:
+      type: object
+      properties: {}
```
Can't we specify this object further?
Like 
```json
{"progress": "float[0-1]", "sim-running": "bool"}
```

antonykamp replied:
I would like to leave it unspecified at the moment like the configuration schemas 😇 I would like to keep the schemas uniformly empty for now 😄

SaturnHafen commented on postman/schemas/schemas/Simulation.yaml:
```diff
@@ -0,0 +1,5 @@
+components:
+  schemas:
+    SimulationConfugration:
```
```suggestion
    SimulationConfiguration:
```

antonykamp replied:
Good catch ⚾ 👀

SaturnHafen commented on postman/schemas/schemas/Token.yaml:
```diff
@@ -0,0 +1,5 @@
+components:
+  schemas:
+    Token:
+      type: object
+      properties: {}
```
You maybe want to specify this further too...

antonykamp replied:
I would like to leave it unspecified at the moment like the configuration schemas 😇 I would like to keep the schemas uniformly empty for now 😄

SaturnHafen commented on postman/schemas/paths/component/data-science-by-id.yaml:
```diff
@@ -0,0 +1,64 @@
+        "204":
+            description: Updated data-science configuration
+        "400":
+            description: Update not possible, you can't update the id
```
When is this the case?

antonykamp replied:
Good question 👍 The first idea was, that you can specify the id by your own. But I'm not a fan of this idea anymore. I was thinking about not allowing to define and update the id. I'll fix that! 🤝 Thank you!

SaturnHafen commented on postman/schemas/paths/component/data-science-result.yaml:
```diff
@@ -0,0 +1,28 @@
+              format: uuid
+    responses:
+        "200":
+            description: Successfull operation
```
I think successful is spelled with one `l`. Might need correction everywhere...

antonykamp replied:
Yes 😇 Thank you very much 🥳

SaturnHafen commented on postman/schemas/paths/component/data-science.yaml:
```diff
@@ -0,0 +1,50 @@
+            description: API token missing
+        "404":
+            description: Simulation not found
+put:
```
Why don't you have to specify the simulation?

antonykamp replied:
You have to specify the simulation object in the configuration objects 😇

SaturnHafen commented on postman/schemas/paths/run-by-id.yaml:
```diff
@@ -0,0 +1,39 @@
+        "404":
+            description: Id not found
+delete:
+    summary: Delete a run and the connected results. If the run is still running, it'll cancel it.
```
```suggestion
    summary: Delete a run and the connected results. If a simulation is still running, cancel it.
```

antonykamp replied:
Good correction 👍 I'll apply that

SaturnHafen commented on postman/schemas/paths/simulation-by-id.yaml:
```diff
@@ -0,0 +1,64 @@
+        "404":
+            description: Id not found
+        
+patch:
```
Just thought about the following:
What happens when I configure & run a simulation, then alter the config and run simulations again. How would you distinguish the runs before and after the config change?

antonykamp replied:
Thanks for pointing out! 😮 That's difficult. My idea is, that starting with the first run, you're not allowed to update the simulation or configurations anymore. 
But we can add a route to copy the whole simulation+component configuration. Then you can change it and run an edited version of the first configuration.

antonykamp replied:
We decided that after executing the simulation, the user is not allowed to edit the configuration. In the future, we'll provide a route to copy a configuration in order to edit an executed one.

SaturnHafen commented on postman/schemas/paths/simulation-by-id.yaml:
```diff
@@ -0,0 +1,64 @@
+        "204":
+            description: Updated simulation configuration
+        "400":
+            description: Update not possible, you can't update the id
```
When does this happen?

antonykamp replied:
see
https://github.com/BP2022-AP1/bp2022-ap1/pull/119/files/e8a2324eaac5d7d2cd8d19081771ea9aa301c0a6#r1150866433

SaturnHafen commented on postman/schemas/paths/run.yaml:
```diff
@@ -0,0 +1,51 @@
+            description: API token missing
+        "404":
+            description: Simulation id not found
+        "409":
+            description: Run already exists with this id
```
Why is the user allowed to specify an id?

antonykamp replied:
See 
https://github.com/BP2022-AP1/bp2022-ap1/pull/119/files/e8a2324eaac5d7d2cd8d19081771ea9aa301c0a6#r1150866433

SaturnHafen reviewed (changes requested):
Wow, thanks a lot for your work!
I have left some comments on the datascience and the auth-token stuff.
Also, would you mind unifying the file endings (newline at EOF vs no newline at EOF)?

CR1337 committed:
fix: typo

CR1337 commented on postman/schemas/paths/component/spawner-by-id.yaml:
```diff
@@ -0,0 +1,64 @@
+                    schema:
+                        $ref: "../../schemas/Spawner.yaml#/components/schema/SpawnerConfiguration"
+        "401":
+            description: API token missing
```
Just for my understanding... how is the API token transmitted?

CR1337 replied:
It is defined in index.yaml

instant-sky commented on postman/schemas/paths/component/fault-injection/platform-blocked-fault-by-id.yaml:
```diff
@@ -0,0 +1,64 @@
+        "404":
+            description: Id not found
+        
+patch:
+    summary: Edit the a platform-blocked-fault configuration.
+    operationId: updatePlatformBlockedFaultConfiguration
```
small typo, should it be "the" platform-blocked-... or "a". This might be in every fault file

antonykamp replied:
Good catch ⚾ Thanks!

instant-sky commented on postman/schemas/paths/component/fault-injection/train-blocked-fault-by-id.yaml:
I think this should be `track-blocked-fault` instead of `train-blocked-fault`.

antonykamp replied:
Thanks for the hint 👍 I replaced it!

instant-sky commented on postman/schemas/paths/component/fault-injection/train-blocked-fault.yaml:
This should be `track-blocked-fault` instead of `train-blocked-fault`

instant-sky commented on postman/schemas/paths/component/fault-injection/train-speed-limit-fault-by-id.yaml:
This should probably be `track-speed-limit-fault` instead of `train-speed-limit-fault`

instant-sky commented on postman/schemas/paths/component/fault-injection/train-speed-limit-fault.yaml:
This should probably be `track-speed-limit-fault` instead of `train-speed-limit-fault`

instant-sky commented on postman/schemas/schemas/FaultInjector.yaml:
```diff
@@ -0,0 +1,20 @@
+        TrainCancelledFaultConfiguration:
+            type: object
+            properties: {}
+        TrainSpeedLimitFaultConfiguration:
```
```suggestion
        TrackSpeedLimitFaultConfiguration:
```

instant-sky commented on postman/schemas/schemas/FaultInjector.yaml:
```diff
@@ -0,0 +1,20 @@
+        TrainSpeedLimitFaultConfiguration:
+            type: object
+            properties: {}
+        TrainBlockedFaultConfiguration:
```
```suggestion
        TrackBlockedFaultConfiguration:
```

instant-sky reviewed (changes requested):
Wow, that's a lot 😮. I mentioned a few small things below, otherwise this looks very promising with my technical knowledge about APIs! 🦜

antonykamp committed:
fix: typo

antonykamp committed:
fix: rename train-blocked-fault to track-blocked-fault

antonykamp committed:
fix: rename TrainBlockedFault to TrackBlockedFault

antonykamp committed:
fix: rename train-speed-limit to track-speed-limit

antonykamp committed:
fix: update routes

Add renaming of train-blocked-fault to track-blocked-fault and train-speed-limit-fault to track-speed-limit-fault

antonykamp committed:
feat: remove token-by-id routes

Remove token-by-id routes to reduce security vulnerabilities and replan later.

instant-sky reviewed (approved):
Looks good now, thank you for the changes! 🎈🎉

antonykamp committed:
fix: remove error responses that you are not allowed to change ids

We wont provide a possibility to change ids. That's why we don't need the error responses either.

CR1337 reviewed (approved):
Looks good. Thanks :)

instant-sky commented on postman/schemas/paths/component/fault-injection/train-cancelled-fault-by-id.yaml:
We agreed, that we want to rename `train-cancelled-fault` to `schedule-blocked-fault`

instant-sky commented on postman/schemas/paths/component/fault-injection/train-cancelled-fault.yaml:
rename `train-cancelled-fault` to `schedule-blocked-fault`

instant-sky reviewed (changes requested):
After the last discussion another small thing appeared:

Lietze reviewed (approved):
The parts of the interlocking controller look good.

antonykamp committed:
feat: add error response when editing an already ran simulation

Already ran simulation cannot be edited.

antonykamp committed:
style: rename train-cancelled-fault to schedule-blocked-fault

antonykamp committed:
fix: remove referencing yaml object in path

instant-sky commented on postman/schemas/index.yaml:
```diff
@@ -0,0 +1,61 @@
+        $ref: './paths/component/fault-injection/train-prio-fault.yaml'
+    /component/fault-injection/train-prio-fault/{id}:
+        $ref: './paths/component/fault-injection/train-prio-fault-by-id.yaml'
+    /component/fault-injection/train-cancelled-fault:
```
I think you missed one `train-cancelled-fault`. Otherwise everything looks good now! 😃

antonykamp replied:
Thanks 👍

antonykamp committed:
fix: replace last train-cancelled-fault with schedule-blocked-fault

antonykamp committed:
feat: merge into one definition file

antonykamp committed:
fix: update internal postman changes

antonykamp commented on postman/schemas/index.yaml:
```diff
@@ -0,0 +1,1238 @@
+security:
+  - ApiKeyAuth: []
+paths:
+  /component/data-science:
```
remove everything from /data-science path

SaturnHafen reviewed (approved):
Why is everything a single file now?

antonykamp committed:
fix: remove data-science paths temporarily

We'll add data-science paths back in a future version.

antonykamp commented:
> Why is everything a single file now?

We need this for code generation in the future.

antonykamp committed:
Merge branch 'dev' into feature/52-implement-first-version-of-rest-api-definition
